### ISSUE TITLE
some crashes seen

### ISSUE THREAD

**Issue by deblauwetom** [paullouisageneau/libdatachannel#671]
This crash:

```
2# __kernel_rt_sigreturn in linux-vdso.so.1
 3# getnameinfo in /lib/libc.so.6
 4# rtc::impl::TcpTransport::prepare(sockaddr const*, unsigned int) in /usr/lib/libdatachannel.so.0.17.7
 5# 0x0000007F92F5193C in /usr/lib/libdatachannel.so.0.17.7
 6# 0x0000007F92F51F1C in /usr/lib/libdatachannel.so.0.17.7
 7# 0x0000007F92F52658 in /usr/lib/libdatachannel.so.0.17.7
 8# rtc::impl::PollService::process(std::vector<pollfd, std::allocator<pollfd> >&) in /usr/lib/libdatachannel.so.0.17.7
 9# rtc::impl::PollService::runLoop() in /usr/lib/libdatachannel.so.0.17.7
10# 0x0000007F951E718C in /usr/lib/libstdc++.so.6
11# 0x0000007F94FE9088 in /lib/libc.so.6
12# 0x0000007F9504DF1C in /lib/libc.so.6
```

and this one:
```
2# __kernel_rt_sigreturn in linux-vdso.so.1
 3# free in /lib/libc.so.6
 4# rtc::impl::Queue<std::shared_ptr<rtc::Message> >::~Queue() in /usr/lib/libdatachannel.so.0.17.7
 5# rtc::impl::TlsTransport::~TlsTransport() in /usr/lib/libdatachannel.so.0.17.7
 6# 0x0000007F75B0EB9C in /usr/lib/libdatachannel.so.0.17.7
 7# 0x0000007F75B0EE64 in /usr/lib/libdatachannel.so.0.17.7
 8# 0x0000007F75A9FE7C in /usr/lib/libdatachannel.so.0.17.7
 9# 0x0000007F77B9DD4C in /lib/libc.so.6
10# void std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*>(std::once_flag&, void (std::__future_base::_State_baseV2::*&&)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*&&, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*&&, bool*&&) in /usr/lib/libdatachannel.so.0.17.7
11# 0x0000007F75B0E980 in /usr/lib/libdatachannel.so.0.17.7
12# rtc::impl::ThreadPool::runOne() in /usr/lib/libdatachannel.so.0.17.7
13# rtc::impl::ThreadPool::run() in /usr/lib/libdatachannel.so.0.17.7
14# 0x0000007F77D9718C in /usr/lib/libstdc++.so.6
15# 0x0000007F77B99088 in /lib/libc.so.6
16# 0x0000007F77BFDF1C in /lib/libc.so.6
```
I never seen these with the older 0.15 branch.

The commit I am using now is 9c4de01fb8dadbf4c2a0cbe2bbef87aeee54bf0f.

**Comment by paullouisageneau** [comment]
Do you have indications to reproduce, by any chance?

**Comment by deblauwetom** [comment]
Hello,
It seems it happens when there is a bad connection or somehow the connection gets closed. I have another indication here of another crash:

```
what(): Connection is not open
 0# toadi::toadi_get_backtrace[abi:cxx11]() in /home/toadi/ToadiApp
 1# toadi::toadi_handle_exception(void*, std::type_info*, bool) in /home/toadi/ToadiApp
 2# __cxa_throw in /home/toadi/ToadiApp
 3# rtc::impl::TcpTransport::send(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 4# rtc::impl::TlsTransport::outgoing(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 5# rtc::impl::TlsTransport::send(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 6# rtc::impl::WsTransport::sendFrame(rtc::impl::WsTransport::Frame const&) in /usr/lib/libdatachannel.so.0.17.8
 7# rtc::impl::WsTransport::close() in /usr/lib/libdatachannel.so.0.17.8
 8# rtc::impl::WsTransport::stop() in /usr/lib/libdatachannel.so.0.17.8
 9# 0x0000007F8447AC04 in /usr/lib/libdatachannel.so.0.17.8
10# 0x0000007F8447AFA4 in /usr/lib/libdatachannel.so.0.17.8
11# 0x0000007F8440BEFC in /usr/lib/libdatachannel.so.0.17.8
12# 0x0000007F864B8D4C in /lib/libc.so.6
13# void std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*>(std::once_flag&, void (std::__future_base::_State_baseV2::*&&)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*&&, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*&&, bool*&&) in /usr/lib/libdatachannel.so.0.17.8
14# 0x0000007F8447AAC0 in /usr/lib/libdatachannel.so.0.17.8
15# rtc::impl::ThreadPool::runOne() in /usr/lib/libdatachannel.so.0.17.8
16# rtc::impl::ThreadPool::run() in /usr/lib/libdatachannel.so.0.17.8
17# 0x0000007F866B218C in /usr/lib/libstdc++.so.6
18# 0x0000007F864B4088 in /lib/libc.so.6
19# 0x0000007F86518F1C in /lib/libc.so.6
```

and shortly 3 milliseconds or so there after came this:

```
type std::runtime_error
what(): Connection is not open
 0# app::app_get_backtrace[abi:cxx11]() in /home/app/MyApp
 1# app::app_handle_exception(void*, std::type_info*, bool) in /home/app/MyApp
 2# __cxa_throw in /home/app/MyApp
 3# rtc::impl::TcpTransport::send(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 4# rtc::impl::TlsTransport::outgoing(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 5# rtc::impl::TlsTransport::send(std::shared_ptr<rtc::Message>) in /usr/lib/libdatachannel.so.0.17.8
 6# rtc::impl::WsTransport::sendFrame(rtc::impl::WsTransport::Frame const&) in /usr/lib/libdatachannel.so.0.17.8
 7# rtc::impl::WsTransport::close() in /usr/lib/libdatachannel.so.0.17.8
 8# rtc::impl::WsTransport::stop() in /usr/lib/libdatachannel.so.0.17.8
 9# 0x0000007F8447AC04 in /usr/lib/libdatachannel.so.0.17.8
10# 0x0000007F8447AFA4 in /usr/lib/libdatachannel.so.0.17.8
11# 0x0000007F8440BEFC in /usr/lib/libdatachannel.so.0.17.8
12# 0x0000007F864B8D4C in /lib/libc.so.6
13# void std::call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*>(std::once_flag&, void (std::__future_base::_State_baseV2::*&&)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*), std::__future_base::_State_baseV2*&&, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*&&, bool*&&) in /usr/lib/libdatachannel.so.0.17.8
14# 0x0000007F8447AAC0 in /usr/lib/libdatachannel.so.0.17.8
15# rtc::impl::ThreadPool::runOne() in /usr/lib/libdatachannel.so.0.17.8
16# rtc::impl::ThreadPool::run() in /usr/lib/libdatachannel.so.0.17.8
17# 0x0000007F866B218C in /usr/lib/libstdc++.so.6
18# 0x0000007F864B4088 in /lib/libc.so.6
19# 0x0000007F86518F1C in /lib/libc.so.6



free(): invalid pointer
Aborted
```

and then it really crashed. I don't really know how to reproduce, but it seems like it happens when the connection gets closed by the driver when it's bad reception of the wifi for example.

**Comment by paullouisageneau** [comment]
The last one should be fixed by https://github.com/paullouisageneau/libdatachannel/pull/673

The first one seems to be caused by incorrect synchronisation of the WebSocket connection process, it'll need a bit of refactoring to fix properly.

**Comment by paullouisageneau** [comment]
The first couple ones should be fixed by https://github.com/paullouisageneau/libdatachannel/pull/677

**Comment by paullouisageneau** [comment]
The fixes are released in `v0.17.9`.
